src/elf/symver.rs

```diff
@@ -0,0 +1,902 @@
+                        shdr.sh_size as usize,
+                        shdr.sh_info as usize, // Encodes the number of ElfVerdef entries.
+                    )
+                } else {
+                    return Ok(None);
+                };
+
+            // Get string table which is used to resolve version strings.
+            let verstr = {
+                // Linked section refers to string table.
+                let shdr_link = shdrs.get(link_idx).ok_or_else(|| {
+                    Error::Malformed(
+                        "Section header of string table for SHT_GNU_VERDEF section not found!"
+                            .into(),
+                    )
+                })?;
+
+                Strtab::parse(
+                    bytes,
+                    shdr_link.sh_offset as usize,
+                    shdr_link.sh_size as usize,
+                    0x0, /* Delimiter */
+                )?
+            };
+
+            // Get a slice of bytes of the `version definition` section content.
+            let bytes: &'a [u8] = bytes.pread_with(offset, size)?;
+
+            Ok(Some(VerdefSection {
+                verstr,
+                bytes,
+                count,
+                ctx,
+            }))
+        }
+
+        /// Get an iterator over the [`Verdef`] entries.
+        #[inline]
+        pub fn iter(&'a self) -> VerdefIter<'a> {
+            self.into_iter()
+        }
+    }
+
+    impl<'a> IntoIterator for &'_ VerdefSection<'a> {
+        type Item = <VerdefIter<'a> as Iterator>::Item;
+        type IntoIter = VerdefIter<'a>;
+
+        #[inline]
+        fn into_iter(self) -> Self::IntoIter {
+            VerdefIter {
+                bytes: self.bytes,
+                count: self.count,
+                index: 0,
+                offset: 0,
+                ctx: self.ctx,
+            }
+        }
+    }
+
+    /// Iterator over the [`Verdef`] entries from the [`SHT_GNU_VERDEF`] section.
+    pub struct VerdefIter<'a> {
+        bytes: &'a [u8],
+        count: usize,
+        index: usize,
+        offset: usize,
+        ctx: container::Ctx,
+    }
+
+    impl<'a> Iterator for VerdefIter<'a> {
+        type Item = Verdef<'a>;
+
+        fn next(&mut self) -> Option<Self::Item> {
+            if self.index >= self.count {
+                None
+            } else {
+                self.index += 1;
+
+                // Safe to unwrap as the length of the byte slice was validated in VerdefSection::parse.
+                let ElfVerdef {
+                    vd_version,
+                    vd_flags,
+                    vd_ndx,
+                    vd_cnt,
+                    vd_hash,
+                    vd_aux,
+                    vd_next,
+                } = self.bytes.pread_with(self.offset, self.ctx.le).unwrap();
+
+                // Get a slice of bytes of the `Verdaux` entries.
+                //
+                // | Verdef | .. | Verdef | Verdaux | Verdaux | .. | Verdef | ..
```

Comment:
I don't think this layout is required. e.g. LSB says:
"The section shall contain an array of `Elfxx_Verdef` structures, as described in Figure 10-1, optionally followed by an array of `Elfxx_Verdaux` structures, as defined in Figure 10-2."
So all of the `Verdaux` may come after all of the `Verdef`. It's best to not make any assumptions about ordering (other than that `vd_next` and `vd_aux` won't point backwards), and I don't think other parsers do.

Same for the `Verneed` structures.

Reply:
Interesting, I did interpret the optionally different here, but you are right. This code should not rely on that layout.